AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,233 @@
+        let thresholdVector = CIVector(x: threshold, y: threshold, z: threshold, w: 1)
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputRVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputGVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputBVector")
+
+        // Keeping alpha channel as is
+        colorMatrixFilter?.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")
+
+        // Biasing the output so that it’s either black or white
+        let biasVector = CIVector(x: -threshold, y: -threshold, z: -threshold, w: 0)
+        colorMatrixFilter?.setValue(biasVector, forKey: "inputBiasVector")
+
+        return colorMatrixFilter?.outputImage
+    }
+
+    // ==================================================
+
+    private func extractCardNumber(from textObservations: [VNRecognizedTextObservation]) -> String? {
+        for observation in textObservations {
+            let candidates = observation.topCandidates(Constants.topCandidates)
+            guard let topCandidate = candidates.first,
+                  topCandidate.confidence > Constants.cardNumberConfidence else { continue }
+
+            debugPrint("candidates >> \(topCandidate.string)")
+
+            let sanitizedCardNumber = topCandidate.string.replacingOccurrences(of: " ", with: "")
+
+            guard sanitizedCardNumber.isOnlyNumbers else { continue }
+            guard sanitizedCardNumber.count >= 13, sanitizedCardNumber.count <= 19 else { continue }
+            guard isValidLuhn(sanitizedCardNumber) else { continue }
+
+            debugPrint("VALID CARD: \(sanitizedCardNumber)")
+
+            return sanitizedCardNumber
+        }
+
+        return nil
+    }
+
+    private func extractExpireDate(from textObservations: [VNRecognizedTextObservation]) -> Date? {
+        for observation in textObservations {
+            let candidates = observation.topCandidates(Constants.topCandidates)
+            guard let topCandidate = candidates.first,
+                  topCandidate.confidence > Constants.expireDateConfidence else { continue }
+
+            let expireDateMatch = extractMatch(from: topCandidate.string, using: Constants.expireDateRegex)
+            guard let expireDateMatch else { continue }
+
+            debugPrint("VALID DATE: \(expireDateMatch)")
+            return date(from: expireDateMatch)
+        }
+
+        return nil
+    }
+
+    private func isValidLuhn(_ number: String) -> Bool {
+        guard number.allSatisfy(\.isNumber) else { return false }
+
+        var sum = 0
+        let reversedDigits = number.reversed().map { Int(String($0))! }
+
+        for (index, digit) in reversedDigits.enumerated() {
+            if index % 2 == 1 {
+                let doubled = digit * 2
+                sum += (doubled > 9) ? (doubled - 9) : doubled
+            } else {
+                sum += digit
+            }
+        }
+
+        return sum % 10 == 0
+    }
```

Comment:
Can we create `String` extension and/or reuse Luhn check from `CardNumberValidator` instead of copy-pasting the code?

Reply:
For now I will let it like this since this is only being used here in the parser. 
Injecting the CardNumberValidador for a luhn check fuction that doesn't change creates more hassle than the benefits (knowing it will never change). Also, no need to have a dependency with `AdyenCard`. `CardNumberValidator` does more than just a luhn check.
We don't validate the card number in the card scanner, that will be done on the card component.

Reply:
`CardNumberValidator` can be moved to `AdyenEncryption` module, since it is useful for both CardComponent and for custom card integration.